docs/src/user-guide/tutorials/auction.md

```diff
@@ -0,0 +1,351 @@
+
+Add the following to the contract to implement the `withdraw` method:
+
+```rust
+pub fn withdraw(mut self, mut ctx: Context) -> bool {
+    let amount: u256 = self.pending_returns[ctx.msg_sender()]
+
+    if amount > 0 {
+        self.pending_returns[ctx.msg_sender()] = 0
+        ctx.send_value(to: ctx.msg_sender(), wei: amount)
+    }
+    return true
+}
+```
+
+>**Note** that in this case `mut` is used with `ctx` because `send_value` is making changes to the blockchain (it is moving ETH from one address to another).
+
+
+### End the auction
+
+Finally, you need to add a way to end the auction. This will check whether the bidding period is over, and if it is, automatically trigger the payment to the beneficiary and emit the address of the winner in an event. 
+
+First, check the auction is not still live - if the auction is live you cannot end it early. If an attempt to end the auction early is made, it should revert using a `AuctionNotYetEnded` struct, which can look as follows:
+
+```rust
+struct AuctionNotYetEnded {
+}
+```
+
+You should also check whether the auction was *already* ended by a previous valid call to this method. In this case, revert with a `AuctionEndAlreadyCalled` struct:
+
+```rust
+struct AuctionEndAlreadyCalled {}
+```
+
+If the auction is still live, you can end it. First set `self.ended` to `true` to update the contract state. Then emit the event using `ctx.emit()`. Then, send the ETH to the beneficiary. Again, the order is important - you should always send value last to protect against re-entrancy.
+Your method can look as follows:
+
+```rust
+pub fn action_end(mut self, mut ctx: Context) {
+    if ctx.block_timestamp() <= self.auction_end_time {
+        revert AuctionNotYetEnded()
+    }
+    if self.ended {
+        revert AuctionEndAlreadyCalled()
+    }
+    self.ended = true
+    ctx.emit(AuctionEnded(winner: self.highest_bidder, amount: self.highest_bid))
+
+    ctx.send_value(to: self.beneficiary, wei: self.highest_bid)
+}
+```
+
+Congratulations! You just wrote an open auction contract in Fe!
+
+
+## View functions
+
+To help test the contract without having to decode transaction logs, you can add some simple functions to the contract that simply report the current values for some key state variables (specifically, `highest_bidder`, `highest_bid` and `ended`). This will allow a user to use `eth_call` to query these values in the contract. `eth_call` is used for functions that do not update the state of the blockchain and costs no gas because the queries can be performed on local data.
+
+You can add the following functions to the contract:
+
+```rust
+pub fn check_highest_bidder(mut self) -> address {
```

Comment:
I didn't spot this in my first review but these do not need to take `mut` as they are readonly